File: backend/app/models/post.py
import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base
from app.models.enums import PostCategory, str_enum


class Post(Base):
    __tablename__ = "posts"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    author_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    author_name: Mapped[str] = mapped_column(String, nullable=False)
    author_avatar_url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    title: Mapped[str] = mapped_column(String, nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    category: Mapped[PostCategory] = mapped_column(str_enum(PostCategory), nullable=False)
    related_resource_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    reaction_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    comment_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    report_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_published: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

File: lib/features/community/data/post.dart
class Post {
  final String id;
  final String authorName;
  final String? authorAvatarUrl;
  final String title;
  final String body;
  final String category;
  final int reactionCount;
  final bool hasReacted;
  final int commentCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Post({
    required this.id,
    required this.authorName,
    required this.authorAvatarUrl,
    required this.title,
    required this.body,
    required this.category,
    required this.reactionCount,
    required this.hasReacted,
    required this.commentCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: json['id'] as String,
      authorName: json['author_name'] as String,
      authorAvatarUrl: json['author_avatar_url'] as String?,
      title: json['title'] as String,
      body: json['body'] as String,
      category: json['category'] as String,
      reactionCount: json['reaction_count'] as int,
      hasReacted: json['has_reacted'] as bool,
      commentCount: json['comment_count'] as int,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Post copyWith({int? reactionCount, bool? hasReacted, int? commentCount}) =>
      Post(
        id: id,
        authorName: authorName,
        authorAvatarUrl: authorAvatarUrl,
        title: title,
        body: body,
        category: category,
        reactionCount: reactionCount ?? this.reactionCount,
        hasReacted: hasReacted ?? this.hasReacted,
        commentCount: commentCount ?? this.commentCount,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}
